File: src/advent_of_code_2019/day_18.clj
(ns advent-of-code-2019.day-18
  (:require [clojure.java.io :as io]
            [clojure.string :as string]
            [clojure.set :as sets]
            [clojure.test :refer :all])
  (:import [java.util HashMap HashSet PriorityQueue]))


(set! *warn-on-reflection* true)


(def ex1 "#########
#b.A.@.a#
#########")


(def ex2 "########################
#f.D.E.e.C.b.A.@.a.B.c.#
######################.#
#d.....................#
########################")


(def ex3 "########################
#...............b.C.D.f#
#.######################
#.....@.a.B.c.d.A.e.F.g#
########################")


(def ex4 "#################
#i.G..c...e..H.p#
########.########
#j.A..b...f..D.o#
########@########
#k.E..a...g..B.n#
########.########
#l.F..d...h..C.m#
#################")

(def ex5 "########################
#@..............ac.GI.b#
###d#e#f################
###A#B#C################
###g#h#i################
########################")


(defn read-cell
  [^Character cell]
  (cond (= cell \#) :wall
        (= cell \.) :space
        (= cell \@) :start
        (Character/isLowerCase cell) (keyword "key" (str cell))
        (Character/isUpperCase cell) (keyword "door" (string/lower-case (str cell)))
        :else (throw (IllegalStateException. (str "Invalid cell: " cell)))))


(defn find-start
  [maze w h]
  (->> (for [y (range h)
             x (range w)]
         [x y])
       (drop-while (fn [[x y]]
                     (not= :start (get-in maze [y x]))))
       (first)))


(defn ns?
  [kw n]
  (and (keyword? kw) (= n (namespace kw))))

(defn key?
  [kw]
  (ns? kw "key"))

(defn door?
  [kw]
  (ns? kw "door"))


(defn find-all-keys
  [maze]
  (into #{}
        (comp (mapcat identity)
              (filter key?))
        maze))


(defn init
  [maze]
  (let [h (count maze)
        w (count (first maze))]
    {:maze maze
     :size [w h]
     :start (find-start maze w h)
     :all-keys (find-all-keys maze)}))


(defn read-input
  ([] (read-input (slurp (io/resource "day_18.txt"))))
  ([input]
   (init (mapv (partial mapv read-cell) (string/split-lines input)))))


(deftest test-key?-and-door?
  (is (key? :key/a))
  (is (not (key? :door/a)))
  (is (not (key? :start)))
  (is (not (key? nil))))


(defn key->door
  [k]
  (keyword "door" (name k)))

(defn door->key 
  [d]
  (keyword "key" (name d)))


(def up [0 -1])
(def down [0 1])
(def right [1 0])
(def left [-1 0])

; got hint from https://medium.com/@werner.altewischer/advent-of-code-day-18-2019-the-real-challenge-aea3d4e96708
; to model the search space as a graph with each node being a tuple of the x,y 
; coordinate of the current position plus the set of keys that have been 
; acquired so far -> [x y key-set] 

(defn adjacent-part1
  [maze [x y ks :as pos]]
  (reduce (fn [adj [dx dy]]
            (let [nx (+ x dx)
                  ny (+ y dy)
                  c (get-in maze [ny nx])]
              (cond ; it's a key, pick it up.
                    (key? c)
                    (conj adj [nx ny (conj ks c)])
                    ; it's a door and we have the key.
                    (and (door? c) (contains? ks (door->key c)))
                    (conj adj [nx ny ks])
                    ; it's a space
                    (#{:start :space} c)
                    (conj adj [nx ny ks])
                    ; off the maze, a wall, or a door we don't have the key for 
                    :else
                    adj)))
          #{} 
          [up down left right]))


(deftest test-adjacent-part1
  (let [{:keys [maze]} (read-input ex1)]
    (testing "can move to spaces but not walls"
      (is (= #{[4 1 #{}] [6 1 #{}]} (adjacent-part1 maze [5 1 #{}]))))
    (testing "moving to a cell with a key puts the key in key set"
      (is (= #{[5 1 #{}] [7 1 #{:key/a}]} (adjacent-part1 maze [6 1 #{}]))))
    (testing "moving to a key which has already been obtained"
      (is (= #{[5 1 #{:key/a}] [7 1 #{:key/a}]} (adjacent-part1 maze [6 1 #{:key/a}]))))
    (testing "can move to cell with door if key is in key set"
      (is (= #{[5 1 #{}]} (adjacent-part1 maze [4 1 #{}])))
      (is (= #{[3 1 #{:key/a}] [5 1 #{:key/a}]} (adjacent-part1 maze [4 1 #{:key/a}]))))))



; https://en.wikipedia.org/wiki/Dijkstra%27s_algorithm#Pseudocode
; 1  function Dijkstra(Graph, source):
; 2
; 3      create vertex set Q
; 4
; 5      for each vertex v in Graph:             
; 6          dist[v] ← INFINITY                  
; 7          prev[v] ← UNDEFINED                 
; 8          add v to Q                      
;10      dist[source] ← 0                        
;11      
;12      while Q is not empty:
;13          u ← vertex in Q with min dist[u]    
;14                                              
;15          remove u from Q 
;16          
;17          for each neighbor v of u:           // only v that are still in Q
;18              alt ← dist[u] + length(u, v)
;19              if alt < dist[v]:               
;20                  dist[v] ← alt 
;21                  prev[v] ← u 
;22
;23      return dist[], prev[]

(defn has-all-keys?
  [pos all-keys]
  (= all-keys (last pos)))


(defn dijkstra-part1
  [{:keys [all-keys maze] [start-x start-y] :start}]
  (let [pq (PriorityQueue. 10000 (fn [x y] (compare (:dist x) (:dist y))))
        all-dist (HashMap.) 
        prev (HashMap.) 
        visited (HashSet.)
        min-end (volatile! nil)]
    (.add pq {:pos [start-x start-y #{}] :dist 0})
    (while (and (not= 0 (.size pq)) (nil? @min-end))
      (let [{u :pos dist :dist} (.poll pq)]
        (if (has-all-keys? u all-keys)
          (vreset! min-end {:dist dist :pos u})
          (do
            (doseq [v (adjacent-part1 maze u)]
              (when (not (.contains visited v))
                (let [alt (inc dist)
                      dist-v (get all-dist v Long/MAX_VALUE)]
                  (when (< alt dist-v)
                    (.remove pq {:pos v :dist dist-v})
                    (.add pq {:pos v :dist alt})
                    (.put all-dist v alt)
                    (.put prev v u)))))
            (.add visited u)))))
    (:dist @min-end)))


(deftest test-part-1
  (is (= 8 (dijkstra-part1 (read-input ex1))))
  (is (= 86 (dijkstra-part1 (read-input ex2))))
  (is (= 132 (dijkstra-part1 (read-input ex3))))
  #_(is (= 136 (dijkstra-part1 (read-input ex4))))
  (is (= 81 (dijkstra-part1 (read-input ex5)))))


#_(time (dijkstra-part1 (read-input))) ; ~ 165s, ~ 2:45min, 5406 steps


(comment
  (let [b1 (doto (java.util.BitSet. 26)
             (.set 25 true)
             (.set 24 true)
             (.set 23 true))
        b2 (doto (java.util.BitSet. 26)
             (.set 25 true)
             (.set 24 true)
             (.set 23 true))
        b3 (doto (java.util.BitSet. 26)
             (.set 25 true))]
    (prn (= b1 b2))
    (prn (= b2 b3))
    (prn (= b1 b3)))
  
  )


; part 2

(defn vec+
  [[x0 y0] [x1 y1]]
  [(+ x0 x1) (+ y0 y1)])


(defn part2-maze
  [{:keys [maze start] :as input}]
  (let [from-start (partial vec+ start)
        new-walls (map from-start [up down left right [0 0]])
        new-starts (into [] 
                         (map from-start)
                         [[-1 -1] [1 -1] [-1 1] [1 1]])]
    (-> (reduce (fn [in [x y]]
                  (assoc-in in [:maze y x] :wall))
                input
                new-walls)
        (dissoc :start)
        (assoc :starts new-starts))))


(def ex6 "#######
#a.#Cd#
##...##
##.@.##
##...##
#cB#Ab#
#######")

(def ex7 "###############
#d.ABC.#.....a#
######...######
######.@.######
######...######
#b.....#.....c#
###############")

(def ex8 "#############
#DcBa.#.GhKl#
#.###...#I###
#e#d#.@.#j#k#
###C#...###J#
#fEbA.#.FgHi#
#############")

(def ex9 "#############
#g#f.D#..h#l#
#F###e#E###.#
#dCba...BcIJ#
#####.@.#####
#nK.L...G...#
#M###N#H###.#
#o#m..#i#jk.#
#############")


(defn render-input
  [{m :maze [w h] :size ss :starts}]
  (doseq [y (range h)]
    (doseq [x (range w)]
      (let [cell (get-in m [y x])]
        (print (cond (contains? ss [x y]) \@
                     (= cell :wall) \#
                     (= cell :space) \.
                     (= cell :start) \@
                     (door? cell) (string/upper-case (name cell))
                     (key? cell) (name cell)))))
    (println)))

#_(= ex6 (string/trim (with-out-str (render-input (read-input ex6)))))
#_(render-input (part2-maze (read-input ex6)))
#_(clojure.pprint/pprint (part2-maze (read-input ex6)))



; using hint from https://medium.com/@werner.altewischer/advent-of-code-day-18-2019-the-real-challenge-aea3d4e96708
; to model each node in the search space graph as a 9 tuple:
; [robot-a-pos robot-b-pos robot-c-pos robot-d-pos key-set]
; [key-set [x0 y0] [x1 y1] [x2 y2] [x3 y3]]

(def ex2b "#######
#a.#Cd#
##@#@##
#######
##@#@##
#cB#Ab#
#######")

#_(let [{:keys [maze starts]} (part2-maze (read-input ex6))]
    (clojure.pprint/pprint
      (adjacent-part2 maze (conj (vec starts) #{}))))
(defn adjacent-part2
  "Returns all possible adjacent nodes from the given node. Valid moves are
  moving any single robot one space. If a key is encountered it is added to the
  key set. A robot can only move onto a door if the key is already in the key
  set. Wrinkle: a robot should in some cases be able to put a key back down to
  be able to back track. This was not noticed in part 1 but the correct answer
  was still found. This function doesn't handle that case."
  [maze [r0 r1 r2 r3 ks :as coord]]
  (reduce (fn [adj [idx dir]]
            (let [[x y :as p] (nth coord idx)
                  [nx ny :as np] (vec+ p dir)
                  c (get-in maze [ny nx])]
              (cond ; it's a key, pick it up.
                    (key? c)
                    (conj adj (-> (assoc coord idx np)
                                  (update 4 conj c)))
                    ; it's a door and we have the key.
                    (and (door? c) (contains? ks (door->key c)))
                    (conj adj (assoc coord idx np))
                    ; it's a space
                    (#{:start :space} c)
                    (conj adj (assoc coord idx np))
                    ; off the maze, a wall, or a door we don't have the key for 
                    :else
                    adj)))
          #{}
          (for [idx (range 4)
                dir [up down left right]]
            [idx dir])))


(deftest test-adjacent-part2
  (let [{maze :maze [r0 r1 r2 r3] :starts} (part2-maze (read-input ex6))
        pos [r0 r1 r2 r3 #{}]]
    (is (= #{[[2 1] r1 r2 r3 #{}]}
           (adjacent-part2 maze pos)))
    (is (= #{[[1 1] r1 r2 r3 #{:key/a}]
             pos}
           (adjacent-part2 maze [[2 1] r1 r2 r3 #{}])))
    (is (= #{[[2 1] r1 r2 r3 #{:key/a}]
             [[1 1] r1 r2 [4 5] #{:key/a}]}
           (adjacent-part2 maze [[1 1] r1 r2 r3 #{:key/a}])))))


(defn dijkstra-part2
  [{:keys [all-keys maze] [r0 r1 r2 r3] :starts}]
  (let [pq (PriorityQueue. 10000 (fn [x y] (compare (:dist x) (:dist y))))
        all-dist (HashMap.) 
        prev (HashMap.) 
        visited (HashSet.)
        min-end (volatile! nil)]
    (.add pq {:pos [r0 r1 r2 r3 #{}] :dist 0})
    (while (not= 0 (.size pq))
      (let [{u :pos dist :dist} (.poll pq)]
        (when (and (has-all-keys? u all-keys)
                   (< dist (:dist @min-end Long/MAX_VALUE)))
          (vreset! min-end {:dist dist :pos u}))
        (doseq [v (adjacent-part2 maze u)]
          (when (not (.contains visited v))
            (let [alt (inc dist)
                  dist-v (get all-dist v Long/MAX_VALUE)]
              (when (< alt dist-v)
                (.remove pq {:pos v :dist dist-v})
                (.add pq {:pos v :dist alt})
                (.put all-dist v alt)
                (.put prev v u)))))
        (.add visited u)))
    (:dist @min-end)))


(deftest test-dijkstra-part2
  #_(is (= 8 (dijkstra-part2 (part2-maze (read-input ex6)))))
  #_(is (= 24 (dijkstra-part2 (part2-maze (read-input ex7)))))
  #_(is (= 32 (dijkstra-part2 (part2-maze (read-input ex8)))))
  #_(is (= 72 (dijkstra-part2 (part2-maze (read-input ex9))))))


#_(time (dijkstra-part2 (part2-maze (read-input))))
#_(.maxMemory (Runtime/getRuntime))

File: src/advent_of_code_2019/day_18.java
package advent_of_code_2019;

import java.util.ArrayList;
import java.util.BitSet;
import java.util.HashMap;
import java.util.HashSet;
import java.util.PriorityQueue;



public class day_18 {
  private static final String ex1 = "#########\n#b.A.@.a#\n#########";
  private static final String ex2 = "########################\n#f.D.E.e.C.b.A.@.a.B.c.#\n######################.#\n#d.....................#\n########################";
  private static final String ex3 = "########################\n#...............b.C.D.f#\n#.######################\n#.....@.a.B.c.d.A.e.F.g#\n########################";
  private static final String ex4 = "#################\n#i.G..c...e..H.p#\n########.########\n#j.A..b...f..D.o#\n########@########\n#k.E..a...g..B.n#\n########.########\n#l.F..d...h..C.m#\n#################";
  private static final String ex5 = "########################\n#@..............ac.GI.b#\n###d#e#f################\n###A#B#C################\n###g#h#i################\n########################";

  public static void main(String[] args) {
    day_18 x = new day_18();
    x.test();
    x.testDijkstraStep();
    System.out.println(x.findAllKeys(ex1));
    System.out.println(x.findAllKeys(ex2));
  }

  public int findAllKeys(String maze) {
    return findAllKeys(parse(maze));
  }

  public void test() {
    testQ();
  }

  public void testQ() {
    // ensure PriorityQueue of State objects works as expected
    PriorityQueue<State> q = new PriorityQueue<>();
    byte width = 0;
    byte height = 0;
    byte keyCount = 2;
    Cell[][] dummyMaze = new Cell[height][width];
    BitSet s0Keys = new BitSet(keyCount);

    State s0 = new State(dummyMaze, new Vector(0, 0), s0Keys, keyCount, width, height, 3);
    State s1 = new State(dummyMaze, new Vector(0, 1), s0Keys, keyCount, width, height, 4);
    State s1b = new State(dummyMaze, new Vector(0, 1), s0Keys, keyCount, width, height, 2);

    // state with min distance is always first

    q.add(s1);
    q.add(s0);

    is(q.poll() == s0);
    is(q.poll() == s1);
    is(q.isEmpty());

    // can update distance by removing and adding new object
    q.add(s0);
    q.add(s1);

    is(q.peek() == s0);
    q.remove(s1);
    q.add(s1b);
    is(q.peek() == s1b);
  }

  public void is(boolean condition) {
    if (!condition) {
      throw new IllegalStateException("condition failed");
    }
  }

//; https://en.wikipedia.org/wiki/Dijkstra%27s_algorithm#Pseudocode
//; 1  function Dijkstra(Graph, source):
//; 2
//; 3      create vertex set Q
//; 4
//; 5      for each vertex v in Graph:             
//; 6          dist[v] ← INFINITY                  
//; 7          prev[v] ← UNDEFINED                 
//; 8          add v to Q                      
//;10      dist[source] ← 0                        
//;11      
//;12      while Q is not empty:
//;13          u ← vertex in Q with min dist[u]    
//;14                                              
//;15          remove u from Q 
//;16          
//;17          for each neighbor v of u:           // only v that are still in Q
//;18              alt ← dist[u] + length(u, v)
//;19              if alt < dist[v]:               
//;20                  dist[v] ← alt 
//;21                  prev[v] ← u 
//;22
//;23      return dist[], prev[]
  int findAllKeys(State start) {
    int dist = -1;

    HashMap<State, Integer> dists = new HashMap<>();
    HashSet<State> visited = new HashSet<>();
    PriorityQueue<State> q = new PriorityQueue<State>(1000000);
    q.add(start);

    while(!q.isEmpty()) {
      int stepResult = dijkstraStep(q, visited, dists);
      if (stepResult != -1) {
        return stepResult;
      }
    }

    return dist;
  }

  int dijkstraStep(PriorityQueue<State> q, HashSet<State> visited, HashMap<State, Integer> dists) {
    State u = q.poll();
    visited.add(u);

    if (u.hasAllKeys()) {
      return u.dist;
    }

    for(State v : u.adjacent()) {
      if(!visited.contains(v)) {
        int alt = u.dist + 1; 
        int currDist = dists.getOrDefault(v, Integer.MAX_VALUE);

        if (currDist == Integer.MAX_VALUE) {
          // v not in q or dists
          q.add(v);
          dists.put(v, alt);
        } else if (alt < currDist) {
          // found a shorter path to v
          q.remove(v);
          q.add(v.withDist(alt));
          dists.put(v, alt);
        } else {
          // found v again but via a longer path
        }
      }
    }

    return -1;
  }

  void testDijkstraStep() {
    HashSet<State> visited = new HashSet<>();

  }

  State parse(String input) {
    String[] lines = input.split("\n");
    byte height = (byte) lines.length;
    byte width = (byte) lines[0].length();
    Vector start = null;
    byte keyCount = 0;
    Cell[][] maze = new Cell[height][width];

    for(int y = 0; y < height; y++) {
      String line = lines[y];
      for(int x = 0; x < width; x++) {
        char c = line.charAt(x);
        Cell cell = null;

        if (c == '#') {
          cell = new Cell(CellType.WALL);
        } else if (c == '.') {
          cell = new Cell(CellType.SPACE);
        } else if (c == '@') {
          cell = new Cell(CellType.START);
          start = new Vector(x, y);
        } else if (Character.isLowerCase(c)) {
          cell = new Cell(CellType.KEY, c);
          keyCount++;
        } else if (Character.isUpperCase(c)) {
          cell = new Cell(CellType.DOOR, c);
        } else {
          throw new IllegalStateException("Invalid cell: " + c);
        }

        maze[y][x] = cell;
      }
    }

    if (keyCount == 0) {
      throw new IllegalStateException("didn't find any keys!");
    }

    if (start == null) {
      throw new IllegalStateException("didn't find start cell");
    }

    BitSet keys = new BitSet(keyCount);
    return new State(maze, start, keys, keyCount, width, height, 0); 
  }

  public enum CellType {
    WALL,
    START,
    SPACE,
    DOOR,
    KEY
  }


  public class Cell {
    public final CellType type;
    public final char name; 

    Cell(CellType type) {
      this(type, '-');
    }

    Cell(CellType type, char name) {
      this.type = type;
      this.name = Character.toLowerCase(name);
    }

    boolean isKey() {
      return CellType.KEY == this.type;
    }

    boolean isDoor() {
      return CellType.DOOR == this.type;
    }

    boolean isSpace() {
      return CellType.START == this.type || CellType.SPACE == this.type;
    }
  }


  public class Vector {
    public final int x;
    public final int y;

    Vector(int x, int y) {
      this.x = x;
      this.y = y;
    }

    Vector move(Vector dir) {
      return new Vector(x + dir.x, y + dir.y);
    }

    boolean inBounds(int width, int height) {
      return x >= 0 && x < width && y >= 0 && y < height;
    }

    @Override
    public boolean equals(Object obj) {
      Vector rhs = (Vector) obj;
      return x == rhs.x && y == rhs.y;
    }
  }


  public class State implements Comparable {
    private final Vector[] dirs = new Vector[]{ new Vector(1,0), new Vector(-1, 0), new Vector(0,1), new Vector(0,-1) };
    public final Cell[][] maze;
    public final Vector pos;
    public final BitSet keys;
    public final byte keyCount;
    public final byte width;
    public final byte height;
    public final int dist;

    State(Cell[][] maze, Vector pos, BitSet keys, byte keyCount, byte width, byte height, int dist) {
      this.maze = maze;
      this.pos = pos;
      this.keys = keys;
      this.keyCount = keyCount;
      this.width = width;
      this.height = height;
      this.dist = dist;
    }

    State addKey(char key) {
      BitSet newKeys = (BitSet) keys.clone();
      newKeys.set(keyIndex(key));
      return new State(maze, pos, newKeys, keyCount, width, height, dist);
    }

    boolean hasAllKeys() {
      return keys.cardinality() == keyCount; 
    }

    boolean haveKey(char name) {
      return keys.get(keyIndex(name)); 
    }

    int keyIndex(char key) {
      return key - 'a';
    }

    ArrayList<State> adjacent() {
      ArrayList<State> adj = new ArrayList<>();
      for (Vector dir : dirs) {
        Vector newPos = pos.move(dir);
        if (newPos.inBounds(width, height)) {
          Cell cell = maze[newPos.y][newPos.x];
          if (cell.isSpace()) {
            adj.add(new State(maze, newPos, keys, keyCount, width, height, dist + 1));
          } else if (cell.isDoor() && haveKey(cell.name)) {
            adj.add(new State(maze, newPos, keys, keyCount, width, height, dist + 1));
          } else if (cell.isKey() && !haveKey(cell.name)) {
            // can move to cell and not pick up key?
            BitSet newKeys = (BitSet) keys.clone();
            newKeys.set(keyIndex(cell.name));
            adj.add(new State(maze, newPos, newKeys, keyCount, width, height, dist + 1));
          }
        }
      }

      return adj;
    }

    State withDist(int newDist) {
      return new State(maze, pos, keys, keyCount, width, height, newDist);
    }

    @Override
    public boolean equals(Object obj) {
      State rhs = (State) obj;
      return keys.equals(rhs.keys) && pos.equals(rhs.pos);
    }

    public int compareTo(Object rhs) {
      int x = this.dist;
      int y = ((State)rhs).dist;

      if (x < y) {
        return -1;
      } else if (y < x) {
        return 1;
      } 

      return 0;
    }
  }
}
